compiler/parser/RdrHsSyn.hs

```diff
@@ -4359,6 +4391,37 @@ data SumOrTuple
   = Sum ConTag Arity (LHsExpr GhcPs)
   | Tuple [LHsTupArg GhcPs]
 
+mkWrittenSumOrTuple :: Boxity -> SrcSpan -> SumOrTuple -> P (HsExpr GhcPs)
+mkWrittenSumOrTuple boxity span sumOrTuple = do
+  isDaml <- getBit DamlSyntaxBit
+  result <- mkSumOrTuple boxity span sumOrTuple
+  let idVar = mkRdrUnqual (mkVarOcc "x")
+      idLam =
+        HsLam noExt
+          (matchGroup noSrcSpan (Match
+            noExt
+            LambdaExpr
+            [VarPat noExt (noLoc idVar)]
+            Nothing
+            (GRHSs
+              noExt
+              [noLoc (GRHS noExt [] (noLoc (HsVar noExt (noLoc idVar))))]
+              (noLoc emptyLocalBinds))))
+  case sumOrTuple of
+    Tuple _ | isDaml -> pure $ HsApp noExt userWrittenTuple (noLoc result)
+    _ -> pure result
+
+unwrapWrittenSumOrTuple :: LHsExpr GhcPs -> Maybe (LHsExpr GhcPs)
+unwrapWrittenSumOrTuple expr
+  | HsApp _ possibleMagic subpat <- unLoc expr
+  , ExprWithTySig _ possibleMagic' _ <- unLoc possibleMagic
+  , HsAppType _ (unLoc -> HsVar _ (unLoc -> funcVar)) (HsWC _ tyArg1) <- unLoc possibleMagic'
+  , HsTyLit _ (HsStrTy _ (unpackFS -> "userWrittenTuple")) <- unLoc tyArg1
+  , funcVar == magicName
+  = Just subpat
```

Comment:
I wonder - out of masochism more than anything else - if we could define this as a bidirectional pattern so the unwrapper is known to be the wrapper's inverse. Not something we need to do now though.